### ISSUE TITLE
Hardware display issues

### ISSUE THREAD

**Issue by eissa** [issalab/mkturk#53]
Photodiode measurements of an alternating black-white square show that even with apparent timely rendering of frames, these are not delivered in a timely fashion to the display hardware as measured by time to first frame (display lag) as well as total number of frames (frame drops). The severity of this issue varies with display configuration and was particularly problematic when using the SuperDisplay app to drive a Pixel 4xl phone as an external display from a PC. The tests run indicated that onboard display or phone as computer not just as display also have hardware display timing issues, but not as severe in nature. Further testing was needed to systematically assess root causes of the problem and whether the browser can precisely sync to display hardware.

**Comment by eissa** [comment]
After further testing with 60Hz rates of alternating black-white squares while rendering images or 3D scenes, mostly on a macbook pro 16" but also on PC and with variety of external monitors...

Factors that can affect hardware display times and frame drops, loosely in order of importance in tests using macbook pro and PC homebrew (can affect absolute lag and/or frame drops, not always, but worth considering these if such issues are noted):

**TL;DR** -- for the most precise, reliable display of software issued frames (fixed delay, no frame drops), set up hardware with mirror displays both running at 60Hz and canvases under 2000 pixels large, have client window in fullscreen mode, and use 2D canvas based display of images rather than online 3D webgl canvas rendering. Importantly, chrome can sync to monitor refreshes (screen flips) with near millisecond precision, so one can come very close to this ideal by following the guidelines below.

- **going through an app to use smartphone as display** introduces insurmountable disconnect between frame flips and actual screen flips (indeed, the mirroring monitor can fare better demonstrating the disconnect)
- not having full focus on the current mkturk tab. If focus is brought to another tab, and the mouse moves there, will see concomitant frame drops on the mkturk tab

- **mirror mode vs. extended display.** Best to mirror the main monitor of interest although extended display will give good performance (on macbook pro, mirror seemed to help further eliminate jitter in absolute lag, frame drops are minimal to none in both cases; on pc, this wasn't fully tested yet, but mirror and extended gave acceptable results in preliminary testing for absolute lag reliability). Even if extended display performance is satisfactory, by running in mirror mode you explicitly avoid the problem of the user taking focus away from the mkturk tab.

- **non-60 Hz frame rate** -- displays these days can run on different frame rates and often faster than 60Hz. However, how they implement these can be a bit proprietary and may not be reflected in the vsync signal available to the browser at least for now (eg, Apple 120 Hz "ProMotion" display). To be safe, all monitors should be run at standard 60Hz. This standard receives the most testing. Even when running at 75Hz on an external monitor, one could see a modified frame pattern (one wide, followed by two narrow) to achieve the slightly elevated frame rate. There is probably some translation being done that is explicitly not being accommodated/tested for in software but can keep an eye on chrome development channels where they are starting to sync to novel frame rates. Also, if you're doing mirror displays, it's just going to be easier to use 60Hz since almost all displays are capable of that rate (mismatching display framerates may be a bad idea for achieving precise, predictable hardware frame timing and just complicates matters)

- **chrome://flags** -- enabling hardware acceleration in chrome settings does make rendering much faster so that instead of using half of a 17ms frame budget, you use only 2-3ms and have plenty of time to spare until the next frame. However, hardware acceleration can cause issues with timing. To get less variance on display lags, you can improve this on mac by setting a few experimental flags. It's worth trying these on whatever system you use to see if it helps reliability (Disable Partial Swap,Disable GPU rasterization,
Disable Zero-copy rasterizer,
Enable Raw Draw,
ANGLE graphics backend = OpenGL)

- **2D canvas vs. webgl canvas** -- both are increasingly reliable if the above hardware setup guidelines are followed along with chrome://flags adjustments. Still 2D canvas has the advantage in terms of precise, reliable first frame timing. There is less jitter, perhaps almost none. For 3D webgl canvas which seems to require the GPU to be more involved, there can appear a delay in displaying the frame sequence. One workaround is to prime the GPU by rendering the first frame offscreen first. This adds a frame delay, but then when shown onscreen, first frame timing reliability can improve. On more powerful systems like a 2080Ti GPU, this priming may be less necessary. Otherwise, both 2D & 3D canvas run with minimal frame drops at reasonably reliable lags and are viable options, especially with a photodiode to backstop the actual vs. desired timing.

- **canvas size** -- for canvases under 2000 x 1500 pixels, display timing is good. Testing larger canvases such as 5K displays may affect timing depending on GPU strength. Good to keep this factor in mind if planning on pushing very challenging rendering scenarios which may start also causing software frame drops (as opposed to hardware frame irregularities being addressed here). However, it's best to simply not push these limits and avoid loading the GPU when you can, pending further stress tests. If one wants to use a very large canvas like 5K x 2K pixels, this may still be doable but may need a stronger system such as one including a 2080Ti GPU or just needs to be re-checked with a photodiode on the current system.

**Comment by eissa** [comment]
Another factor to consider is the **_number of GPUs and number of lanes per GPU_**. For example, a 2080Ti sharing lanes with the neuropix data acq card means the GPU only has 8 lanes. After removing the neuropix data acq card, the 2080Ti had access to the 16x lanes it needs, and display timing seemed to become even tighter (2.5ms jitter instead of 5ms). This inference is purely empirical but the only thing that meaningfully changed was the number of lanes.
Alternatively, we had tried putting 2x GPUs which could also help (unfortunately, the consumer grade motherboard would not even boot). But then that advantage of two cards is offset by the drop in number of lanes per GPU, and reading online, it's complicated to know how much of a gain you will get from more than one GPU.
So for now, we went with one GPU with 16 lanes. In general, this just highlights that one can try to optimize hardware but the benefits will be hard to exactly predict until you test empirically on your setting.

**Comment by eissa** [comment]
Interrupts of screen from other programs (or user input) are always an issue as described in psychtoolbox and can be mitigated by prioritizing your app in windows task manager (can't find this blurb in the psychtoolbox manual at the moment but it is mentioned online). However, behavior is not guaranteed for all possible distracting user interactions + parallel programs/OS. the safest bet is to run your app in full screen and limit use of other programs for millisecond precision.

For chrome, disable partial swap, so that if there is an interrupt, the display latency returns to what it was before interrupt. Currently, this is the only chrome://flag that we set to get good overall graphics performance in testing. Other flags are left as default including hardware acceleration (e.g., raw draw, zero copy, GPU rasterization). The choice of ANGLE Graphics backend determines the display lag (e.g. webgl vs metal on Mac) and using the default or OS specific graphics backend rather than webgl seems to give the shorter latency.

[What is partial swap in chrome://flags?](https://qr.ae/pvfZx0)

**Comment by eissa** [comment]
There are an increasing number of papers showing that recent javascript performance is starting to approach that of psychtoolbox under certain settings. This is just one example:
[Implementation of stimuli with millisecond timing accuracy in online experiments 2020](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0235249)
<img width="1024" alt="Screenshot 2023-01-07 at 4 52 04 PM" src="https://user-images.githubusercontent.com/1245902/211171615-6c5026cc-43eb-4898-b8b1-f461bf44dd18.png">
